### PR TITLE
Create CallOp for KrnlDialect

### PR DESCRIPTION
**Pull request by chentong319**
We defined a lot of Krnl ops for just call a function. There are quite a lot redundant code for lowering from onnx to krnl, and from Krnl to llvm.  
This PR is trying to introduce one CallOp in Krnl to replace those ops.  Changes are
1.  Definition of KrnlCallOp
2. Special builder to help lowering onnx op to krnlCallOp
3. Example of using KrnlCallOp in ResizeOp
4. Test case

Future work: 
1. Lower KrnlCallOp to LLVM
2. Integrated with  runtime api and llvm math calls
3. Replace the existing Krnl ops that is intended for a function call.

### CODE REVIEW

File: src/Conversion/ONNXToKrnl/Tensor/Resize.cpp [hunk 1]
```diff
@@ -119,6 +119,14 @@ struct ONNXResizeOpLowering : public ConversionPattern {
           rewriter, op, memRefType, loc, outputDims, insertDealloc);
     }
 
+    // Call external function when the mode is not "nearest"
+    // Create KrnlCallOp and replace the du chain
+    if (resizeOp.mode() != "nearest") {
+      Value resizeCall = rewriter.create<KrnlCallOp>(loc, alloc, op, operands);
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Conversion/ONNXToKrnl/Tensor/Resize.cpp:125]: let's wait until the builder interface becomes mature.

File: src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp [hunk 2]
```diff
@@ -280,6 +280,22 @@ static FlatSymbolRefAttr getOrInsertUnaryMathFunction(PatternRewriter &rewriter,
...
 }
 
+//===----------------------------------------------------------------------===//
+// KRNL to LLVM: KrnlCallOpLowering
+//===----------------------------------------------------------------------===//
+
+class KrnlCallOpLowering : public ConversionPattern {
+public:
+  explicit KrnlCallOpLowering(MLIRContext *context)
+      : ConversionPattern(KrnlCallOp::getOperationName(), 1, context) {}
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp:290]: Updated

File: src/Dialect/Krnl/KrnlOps.cpp [hunk 3]
```diff
@@ -73,6 +73,52 @@ void KrnlOpsDialect::printType(Type type, DialectAsmPrinter &os) const {
...
+// KrnlCallOp
+//===----------------------------------------------------------------------===//
+
+void KrnlCallOp::build(OpBuilder &builder, ::mlir::OperationState &odsState,
+    Value resultVal, Operation *op, ValueRange operands) {
+  StringRef name = op->getName().getStringRef();
+  ShapedType resultType = resultVal.getType().cast<ShapedType>();
+  Type elementType = resultType.getElementType();
+  std::string funcNameStr;
+  if (elementType.isF32()) {
```
[the remark this answers is not in the document]
Comment (etiotto) [reply src/Dialect/Krnl/KrnlOps.cpp]: ok

File: src/Dialect/Krnl/Krnl.td [hunk 4]
```diff
@@ -34,6 +34,32 @@ def StringType : Type<CPred<"$_self.isa<::mlir::StringType>()">, "string type">;
...
+  }];
+
+  let arguments = (ins StrAttr:$funcName, AnyType:$alloc, Variadic<AnyType>:$parameters);
+  let results =(outs AnyType:$result);
+
+  // builders to build KrnlCallOp from op and operands, helping conversion from onnx to krnl.
+  // The name of function can be determined by the op name and elemnt type of the return,
+  // or given to builder if the simple rule does not work
+  // Attributes of the op will be propagated to KrnlCallOp if the copyAttrs is true
+  let builders = [OpBuilder<(ins "mlir::Value":$alloc, "mlir::Operation *":$op, "mlir::ValueRange":$operands, "bool":$copyAttrs)>,
```
Comment (tungld) [inline src/Dialect/Krnl/Krnl.td:59]: Would it be more generic if we don't associate KrnlCallop with an operation (ONNX operation I guest). For example, I just want to insert a call to a print function, not associate with any ONNX operation.

File: src/Dialect/Krnl/KrnlOps.cpp [hunk 5]
```diff
@@ -73,6 +73,54 @@ void KrnlOpsDialect::printType(Type type, DialectAsmPrinter &os) const {
...
+
+static std::string typeToString(Type ty) {
+  std::string str;
+  llvm::raw_string_ostream out(str);
+  ty.print(out);
+  return out.str();
+}
+
+void KrnlCallOp::build(OpBuilder &builder, ::mlir::OperationState &odsState,
+    std::string funcNameStr, Value resultVal, Operation *op,
```
[the remark this answers is not in the document]
Comment (chentong319) [reply src/Dialect/Krnl/KrnlOps.cpp:88]: I think this check will be done using the runtime api when KrnlCallOp is lowered to llvm.

File: src/Dialect/Krnl/Krnl.td (hunk 4 above)
Comment (tungld) [reply src/Dialect/Krnl/Krnl.td:59]: I see. Thanks!

**Review by tungld — APPROVED** [review]
LGTM! 

`KrnlToLLVM.cpp` was replaced by `ConvertKrnlToLLVM.cpp`.  I think you need to move your code to the new location and create a new KrnlCall.cpp for lowering KrnlCallOp.

**Pull request merged**